core/types/calls.cc

```diff
@@ -3072,6 +3072,59 @@ class Module_tripleEq : public IntrinsicMethod {
     }
 } Module_tripleEq;
 
+// Returns true if the referenced class is a child of T::Enum.
+static bool isEnumClass(const GlobalState &gs, ClassOrModuleRef klass) {
+    if (!klass.exists()) {
+        return false;
+    }
+
+    auto superClass = klass.data(gs)->superClass();
+    return superClass.exists() && superClass == Symbols::T_Enum();
+}
+
+// Returns true if the referenced class is a grandchild of T::Enum, i.e., one of the singleton classes for the enum
+// values.
+static bool isEnumValue(const GlobalState &gs, ClassOrModuleRef klass) {
+    if (!klass.exists()) {
+        return false;
+    }
+
+    auto superClass = klass.data(gs)->superClass();
+    return isEnumClass(gs, superClass);
+}
+
+class T_Enum_tripleEq : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        if (args.args.size() != 1) {
+            return;
+        }
+        auto rhs = args.args[0]->type;
+        if (rhs.isUntyped()) {
+            res.returnType = rhs;
+            return;
+        }
+        auto lhs = args.thisType;
+        ENFORCE(!lhs.isUntyped(), "lhs of T::Enum.=== must be typed");
+        if (Types::glb(gs, rhs, lhs).isBottom() && Types::glb(gs, rhs, Types::String()).isBottom()) {
+            res.returnType = Types::falseClass();
+            return;
+        }
+
+        // If lhs and rhs are both grandchildren of T::Enum, we can safely assume they are singleton classes created by
+        // the T::Enum rewriter, since inheriting from children of T::Enum is otherwise prohibited.
+        bool must_exist = false;
+        auto lhs_unwrapped = unwrapSymbol(gs, lhs, must_exist);
+        auto rhs_unwrapped = unwrapSymbol(gs, rhs, must_exist);
+        if (Types::isSubType(gs, rhs, lhs) && isEnumValue(gs, lhs_unwrapped) && isEnumValue(gs, rhs_unwrapped)) {
```

Comment:
```suggestion
        if (Types::isSubType(gs, rhs, lhs) && isEnumValueClass(gs, lhs_unwrapped) && isEnumValueClass(gs, rhs_unwrapped)) {
```

Reply:
What do you think about reordering this conditional to check if rhs is a subtype of lhs last?

Reply:
What about rewriting this to check that RHS is a subtype of LHS, and LHS is a ClassType with a TEnum name? (You can use the `NameRef::isTEnumName` helper to test that.)